File: git_clones.py
from argparse import ArgumentParser
from re import findall
from subprocess import call
from sys import exit
from urllib.request import urlopen


"""
rootVIII
Download/clone all of a user's PUBLIC source repositories.
Pass the Github user's username with the -u option.
"""


class GitClones:
    def __init__(self, user):
        self.url = f'https://github.com/{user}'
        self.url += '?&tab=repositories&q=&type=source'
        self.git_clone = 'git clone https://github.com/%s/%%s.git' % user
        self.user = user

    def http_get(self):
        return urlopen(self.url).read().decode('utf-8')

    def get_repo_data(self):
        pattern = r"<a\s?href\W+%s/(.*)\"\s+" % self.user
        for line in findall(pattern, self.http_get()):
            yield line.split('\"')[0]

    def get_repos(self):
        return [repo for repo in self.get_repo_data()]

    def download(self):
        _ = [call((self.git_clone % git).split()) for git in self.get_repos()]


if __name__ == "__main__":
    message = 'Usage: python3 git_clones.py -u <github username>'
    parser = ArgumentParser(description=message)
    parser.add_argument('-u', '--user', required=True, help='Github Username')

    clones = GitClones(parser.parse_args().user)
    try:
        clones.download()
    except Exception as err:
        print(f'{type(err).__name__}: {str(err)}')
        exit(1)

File: gitclones.go
package main

// rootVIII gitclones.go

import (
	"flag"
	"fmt"
	"io/ioutil"
	"net/http"
	"os"
	"os/exec"
	"regexp"
	"strings"
)

type clone interface {
	httpReq() string
	findRepositories(requestData string)
	download(repository string)
	setURL()
}

type gitClone struct {
	baseURL  string
	userURL  string
	username string
}

func (gc gitClone) httpReq() string {
	client := &http.Client{}
	req, err := http.NewRequest("GET", gc.userURL, nil)
	resp, err := client.Do(req)
	if err != nil {
		fmt.Printf("request failed: %s\n", gc.userURL)
		os.Exit(1)
	}
	rText, err := ioutil.ReadAll(resp.Body)
	return string(rText)
}

func (gc *gitClone) setURL() {
	gc.userURL = gc.baseURL + gc.username
	gc.userURL += "?&tab=repositories&q=&type=source"
}

func (gc *gitClone) download(repository string) {
	projectURL := gc.baseURL + gc.username
	projectURL += fmt.Sprintf("/%s.git", repository)
	fmt.Printf("Cloning %s\n", repository)
	exec.Command("git", "clone", projectURL).Output()
}

func (gc *gitClone) findRepositories(requestData string) {
	pattern := fmt.Sprintf(`<a\s?href\W+%s\S+`, gc.username)
	result := regexp.MustCompile(pattern)
	links := result.FindAllString(requestData, -1)
	for _, tag := range links {
		gc.download(strings.Split(tag[:len(tag)-1], "/")[2])
	}
}

func main() {
	user := flag.String("u", "", "Provide a valid Github username")
	flag.Parse()
	if len(*user) < 1 {
		fmt.Println("required argument: -u <Github username>")
		os.Exit(1)
	}
	var gitclone clone
	gitclone = &gitClone{username: *user, baseURL: "https://github.com/"}
	gitclone.setURL()
	pageData := gitclone.httpReq()
	gitclone.findRepositories(pageData)
}
